### PR TITLE
Merge pull request #51

### PR DESCRIPTION
**Pull request by qindazhu**
This PR is totally from Dan's PR https://github.com/danpovey/k2/pull/51 about `Array2`  , I just fix a few style issues.

### CODE REVIEW

**Comment by danpovey** [comment]
dont have time to check right now so you guys merge if you think it's OK

**Comment by qindazhu** [comment]
Merging as previous comments in pull request https://github.com/danpovey/k2/pull/51 have been fixed. We can fix any other issues later while we use `Array` in other FSA algorithms.

File: notes/array.txt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    @property
+    def shape(self):
+        # e.g. if indexes.shape is (15,) and
+        # data.shape is (150) -> this.shape would be (15,None)
+        # If data.shape is (150,4), this.shape would be (15,4)
+        # If data.shape is (150,None) (since data is an Array), this.shape
+        #     would be (150,None,None).
+        # The Nones are for dimensions where the shape is not known
+        # because it is variable.
+        return (indexes.shape[0] - 1, None, *data.shape[1:])
```
Comment (csukuangfj) [inline notes/array.txt:38]: The comments are confusing compared to  
`return (indexes.shape[0] - 1, None, *data.shape[1:])`.

Could you please make it clearer? @danpovey

Comment (danpovey) [reply notes/array.txt:38]: I think indexes.shape should be 16 here.  Can you please edit though?  Sorry I am a bit busy today.  I am hoping @qindazhu can have a look at your PR.  Thanks so much!!

**Pull request merged**